subsys/net/lib/fota_download/src/fota_download.c

```diff
@@ -30,59 +29,77 @@ static int download_client_callback(const struct download_client_evt *event)
 
 	switch (event->id) {
 	case DOWNLOAD_CLIENT_EVT_FRAGMENT: {
-		size_t size;
+		size_t file_size;
 
-		err = download_client_file_size_get(&dfu, &size);
+		err = download_client_file_size_get(&dlc, &file_size);
 		if (err != 0) {
-			LOG_ERR("download_client_file_size_get error %d", err);
+			LOG_DBG("download_client_file_size_get err: %d", err);
 			callback(FOTA_DOWNLOAD_EVT_ERROR);
 			return err;
 		}
-		if (size > PM_MCUBOOT_SECONDARY_SIZE) {
+
+		if (file_size > PM_MCUBOOT_SECONDARY_SIZE) {
 			LOG_ERR("Requested file too big to fit in flash\n");
 			callback(FOTA_DOWNLOAD_EVT_ERROR);
 			return -EFBIG;
 		}
 
-		err = flash_img_buffered_write(&flash_img,
-				(u8_t *)event->fragment.buf,
-				event->fragment.len, false);
+		if (first_fragment) {
+			first_fragment = false;
+			int img_type = dfu_target_img_type(event->fragment.buf,
+							event->fragment.len);
+
+			err = dfu_target_init(img_type);
+			if ((err < 0) && (err != -EBUSY)) {
+				LOG_ERR("dfu_target_init error %d", err);
+				return err;
+			}
+
+			err = dfu_target_offset_get(&offset);
+			LOG_INF("Offset: 0x%x", offset);
+
+			if (offset != 0) {
+				/* Abort current download procedure, and
+				 * schedule new download from offset.
+				 */
+				k_delayed_work_submit(&restart_dlc_work,
+						K_SECONDS(1));
+				LOG_INF("Refuse fragment, restart with offset");
+
+				return -1;
+			}
+		}
+
+		err = dfu_target_write(event->fragment.buf,
+				       event->fragment.len);
 		if (err != 0) {
-			LOG_ERR("flash_img_buffered_write error %d", err);
+			LOG_ERR("dfu_target_write error %d", err);
+			err = download_client_disconnect(&dlc);
 			callback(FOTA_DOWNLOAD_EVT_ERROR);
 			return err;
 		}
-		break;
+	break;
 	}
 
 	case DOWNLOAD_CLIENT_EVT_DONE:
-		/* Write with 0 length to flush the write operation to flash. */
-		err = flash_img_buffered_write(&flash_img,
-				(u8_t *)event->fragment.buf,
-				0, true);
+		err = dfu_target_done();
 		if (err != 0) {
-			LOG_ERR("flash_img_buffered_write error %d", err);
+			LOG_ERR("dfu_target_done error: %d", err);
 			callback(FOTA_DOWNLOAD_EVT_ERROR);
 			return err;
 		}
 
-		err = boot_request_upgrade(BOOT_UPGRADE_TEST);
-		if (err != 0) {
-			LOG_ERR("boot_request_upgrade error %d", err);
-			callback(FOTA_DOWNLOAD_EVT_ERROR);
-			return err;
-		}
-		err = download_client_disconnect(&dfu);
+		err = download_client_disconnect(&dlc);
 		if (err != 0) {
-			LOG_ERR("download_client_disconncet error %d", err);
 			callback(FOTA_DOWNLOAD_EVT_ERROR);
 			return err;
 		}
-		callback(FOTA_DOWNLOAD_EVT_FINISHED);
+		first_fragment = true;
 		break;
 
 	case DOWNLOAD_CLIENT_EVT_ERROR: {
```

Comment:
Is `dfu_target` able to be restarted in this case? I guess `dfu_target_init()` will give `-EBUSY`.

Reply:
I'm confused to what this refers to this is checked in `fota_download.c:23`

Reply:
If a `DOWNLOAD_CLIENT_EVT_ERROR` happens, and `fota_download_start()` is called again, how does dfu_target handle that? It looks like `dfu_target_init()` will return `-EBUSY` (because `dfu_target_init()` is called without a `dfu_target_done()` first), failing `fota_download_start()`.

Or does `DOWNLOAD_CLIENT_EVT_DONE` always happen if `DOWNLOAD_CLIENT_EVT_ERROR` happens?

Reply:
Your correct you will not be able to initialize it with a **new target**. So if you call `fota_download_start `with 1 file then if it fails you call it with a 2nd file which is a **different target** you'll get into a pretty weird state where you either write a modem delta to mcuboot or pass a mcuboot update to the modem(which will fail)